### PR TITLE
modules/output: produce empty files by default

### PR DESCRIPTION
**Pull request by stasjok**
This PR removes empty

```lua
vim.cmd([[

]])
```

block from config.

I know that this is already handled by #1889, but I've added tests to make sure that this is the behavior. Also I added the test to make sure that all extraConfigs are outputted.

This also changes various places in other modules with the intention of not setting `extraConfig*` options to empty strings (more specifically whitespace only strings).

NOTE: The tests will fail until #1906, #1907 and #1908 are merged.

### CODE REVIEW

**Comment by MattSturgeon** [comment]
Thanks! I'd already looked at doing this in https://github.com/nix-community/nixvim/pull/1889, however since that one is larger in scope and unfinished it's better to have this merged sooner than later!

It's also good to have tests, which I hadn't implemented in https://github.com/nix-community/nixvim/pull/1889.

Would you mind looking over how I'd done it in that other PR to see if there's any elements you'd like to incorporate in this one?

E.g. it probably makes sense to move `hasContent` into our helpers lib and maybe come up with some kinda wrapper function to reduce boilerplate when wrapping potentially empty lua/vim strings?

Not too important as those things may expand the size/scope of this PR unnecessarily making it harder to review. But worth a quick look.

**Comment by stasjok** [comment]
> Would you mind looking over how I'd done it in that other PR to see if there's any elements you'd like to incorporate in this one?

Your implementation is better, because it doesn't output empty lines. I like it more. But I intentionally didn't rip it off, because I wanted something simple to get started (and functional) + tests. And since you already implemented a better version, I thought that you'll easily rebase to it later. But if you think that it's better to include your version here, I probably can extract it from there.

**Comment by MattSturgeon** [comment]
> But if you think that it's better to include your version here, I probably can extract it from there.

I don't mind too much either way. There's certainly value in keeping this PR simple as you said.

File: lib/neovim-plugin.nix
Comment (MattSturgeon) [inline lib/neovim-plugin.nix:114]: This is intentionally not using mkIf

File: modules/autocmd.nix
Comment (MattSturgeon) [inline modules/autocmd.nix:45]: Unrelated fix

File: modules/highlights.nix
Comment (MattSturgeon) [inline modules/highlights.nix:41]: Is using `mkIf` really so important that we must also use it when concatenating strings?

IMO concatenating optional strings is fine here and in similar cases, @traxys what do you think?

File: modules/opts.nix
Comment (MattSturgeon) [inline modules/opts.nix:60]: Is this a new helper? `concatNonEmptyLines` would be more concise.

File: modules/top-level/output.nix
Comment (MattSturgeon) [inline modules/top-level/output.nix:143]: `extraPlugins` lower down in this file can also be a `mkIf`

**Review by MattSturgeon — CHANGES_REQUESTED** [review]
(Just reviewing the treewide mkIf commit)

- This would be better off in a dedicated PR, IMO.
  I don't see how it relates to this one?
- There's a typo in the commit message (`treeside` -> `treewide`)
- Some stuff is either deliberately not using mkIf or is less readable, especially string concatenation.

**Comment by stasjok** [comment]
> I don't mind too much either way. There's certainly value in keeping this PR simple as you said.

I rewrote it completely. See if it's OK.

Also I replaced a couple of global instances of `optionalString` with `mkIf`. It's in a separate commit, I can drop it if needed.

Before:

```
nix-repl> c.options.extraConfigLuaPre.definitions      
[ "\n\n\n\n" "" "" "-- Ignore the user lua configuration\nvim.opt.runtimepath:remove(vim.fn.stdpath('config'))         
     -- ~/.config/nvim\nvim.opt.runtimepath:remove(vim.fn.stdpath('config') .. \"/after\")  -- ~/.config/nvim/after\nvi
m.opt.runtimepath:remove(vim.fn.stdpath('data') .. \"/site\")     -- ~/.local/share/nvim/site\n" ]

nix-repl> c.options.extraConfigLuaPost.definitions
[ "" "" ]
```

After:

```
nix-repl> c.options.extraConfigLuaPre.definitions  
[ "-- Ignore the user lua configuration\nvim.opt.runtimepath:remove(vim.fn.stdpath('config'))              -- ~/.config
/nvim\nvim.opt.runtimepath:remove(vim.fn.stdpath('config') .. \"/after\")  -- ~/.config/nvim/after\nvim.opt.runtimepath
:remove(vim.fn.stdpath('data') .. \"/site\")     -- ~/.local/share/nvim/site\n" ]

nix-repl> c.options.extraConfigLuaPost.definitions
[ "" ]
```

Notice that there is no empty definitions by default.

But similar cases with empty definitions are still present in some of the plugins, although they are guarded by `mkIf cfg.enable`, if plugin is enabled, they appear. But I definitely not going to edit them now.

File: modules/highlights.nix
Comment (stasjok) [reply modules/highlights.nix:41]: There was two way to do it:

1) Concat strings in a variable, then in mkIf check if it empty.
2) Separate with mkMerge.

I can't decide what's best, I though that this way looks better than when two options are concated with variables and third as is. This way all three options looks the similar.

File: lib/neovim-plugin.nix
Comment (stasjok) [reply lib/neovim-plugin.nix:114]: Why? If `callSetup` is false, plugin produces empty an string value. Later on all this empty strings are concatenated with newlines. If it can be fixed in other places, why not here?

File: modules/autocmd.nix
Comment (stasjok) [reply modules/autocmd.nix:45]: All this fixes was to remove empty extra empty lines from output. This fixes a condition that is guarding `extraConfigLuaPost` from producing empty string.

File: tests/test-sources/modules/output.nix
Comment (MattSturgeon) [inline tests/test-sources/modules/output.nix:22]: I think `name: value:` or `n: v:` is better, to avoid confusion with the overloaded `config` from the module args.

Comment (MattSturgeon) [reply tests/test-sources/modules/output.nix:22]: Also lets name this something like `mkAssertion` or `assertFileContent` or `mkFileAssertion` (e.g.)

File: modules/top-level/output.nix
Comment (stasjok) [reply modules/top-level/output.nix:143]: OK, I can change it too, but this PR is about cleaning up generated configs (mostly removing extra empty lines).

**Review by MattSturgeon — CHANGES_REQUESTED** [review]
(This review is for the other (non-mkIf) changes)

Looks good, mostly just minor nits!

**Comment by stasjok** [comment]
I'll check comments tomorrow. Please decide if a second commit removing empty lines from outputs in a various places is wanted or not.

File: modules/highlights.nix
Comment (MattSturgeon) [reply modules/highlights.nix:41]: Ah I see, so you do this because all parts of the string are optional, so you're avoiding an empty definition?

That makes more sense. Does that apply to all changes though?

File: modules/autocmd.nix
Comment (MattSturgeon) [reply modules/autocmd.nix:45]: Got it.

It is unrelated to the "use mkIf instead of optionalString" commit, but not unrelated to this PR if it solves the bigger issue.

File: lib/neovim-plugin.nix
Comment (MattSturgeon) [reply lib/neovim-plugin.nix:114]: As a general rule, these general use helpers don't know _why_ a feature such as `callSetup` has been disabled, it may be because the caller doesn't want something that feature uses to be evaluated (which mkIf won't solve).

To clarify, `mkIf` still evaluates the content, even when the condition is false.

In this specific case, we'd be calling `toLuaObject cfg.settings` unnecessarily, although maybe that's not a big deal.

But here I think it is better to use `optional`, `optionals`, `optionalAttrs`, `optionalString` or `if then else` as appropriate.

File: modules/top-level/output.nix
Comment (MattSturgeon) [reply modules/top-level/output.nix:143]: When I read "treewide: optionalString -> mkIf" I didn't understand the motivation for the change and just assumed you were doing general cleanup.

**Comment by MattSturgeon** [comment]
> This would be better off in a dedicated PR, IMO.
> I don't see how it relates to this one?

After reading your replies I agree it makes sense in this PR, but it would benefit from a commit message that explains the motivation behind the change, as I was clearly confused 😉

File: lib/neovim-plugin.nix
Comment (stasjok) [reply lib/neovim-plugin.nix:114]: > To clarify, mkIf still evaluates the content, even when the condition is false.

OK. Got it. Since `mkMerge` is already used here, what do you thin about:

```nix
(optionalAttrs callSetup {
   ${extraConfigNamespace} = ''
       require('${luaName}')${setup}(${optionalString (cfg ? settings) (toLuaObject cfg.settings)})
    '';
})
```

I think it preserve current behavior and avoid empty string.

Comment (MattSturgeon) [reply lib/neovim-plugin.nix:114]: Agreed, I think that's the right approach 😃

File: lib/utils.nix
Comment (stasjok) [reply lib/utils.nix]: That was a hard one. `concatNonEmptyLines` was one of the choices, but it's similar to `concatLines` from nixpkgs library and their behavior is different. concatLines is for concatenating one-line-strings, that's why it appends newlines to every string. This function does a different thing. I named it after concatStringsSep with a Sep suffix changed to the actual separator, since my function doesn't separator as an argument.

But since you suggested `concatNonEmptyLines` as a best choice, so be it. I don't like my choice anyway.

File: lib/utils.nix
Comment (stasjok) [reply lib/utils.nix:137]: That's a valid point. To be honest I didn't even considered a possibility that `]]` can be in the string...

**Comment by stasjok** [comment]
Since other PRs are merged, tests are passing.

File: lib/utils.nix
Comment (MattSturgeon) [reply lib/utils.nix]: ~~[This looks like](https://noogle.dev/f/lib/concatLines) the same behaviour to me:~~

```nix
concatLines [ "foo" "bar" ]
=> "foo\nbar\n"
```

`concatLines` is _almost_ the same as `concatStringsSep "\n"`.

Other than the trailing newline, which is only added by `concatLines`, incidentally that's probably what we actually _want_ in this case?

Maybe the implementation should actually be `lines: concatLines (filter hasContent lines)`?

Comment (stasjok) [reply lib/utils.nix]: > Maybe the implementation should actually be `lines: concatLines (filter hasContent lines)`?

No, actually I don't want a trailing newline. The reason for this is that most of the time config parts are done like this:

```nix
{
  var = ''
    text
  ''
}
```
This adds a trailing newline by itself. With `concatLines` it'll be two trailing newlines. In my config I just make sure, that all my parts have trailing newlines by itself (parts from files also have trailing newlines, because I don't use `: set nofixendofline`. But the resulting concatenated result wouldn't be right if some strings are assigned with `var = "text";`. To fix this case I usually use a map with first stripping trailing newlines, than adding exactly one.

File: tests/test-sources/modules/output.nix
Comment (MattSturgeon) [inline tests/test-sources/modules/output.nix:4]: I think we can remove this whole section and access these (where needed) via module args in each test case's module

E.g.

```nix
all-configs.module = 
  {
    config,
    pkgs,
    lib,
    helpers,
    ...
  }:
```

Comment (stasjok) [reply tests/test-sources/modules/output.nix:4]: I wanted to ask something unrelated. Is it intended that I can't just use module as a function without `module` attr? It's because here

https://github.com/nix-community/nixvim/blob/6dc0bda459bcfb2a38cf7b6ed1d6a5d6a8105f00/tests/default.nix#L47

it doesn't handled that it can be a function. To workaround this I have to do `all-configs.module` instead of just `all-configs`

File: lib/utils.nix
Comment (MattSturgeon) [reply lib/utils.nix]: I'm not sure I agree, since many of the places changed in _this_ PR are going from either using the pattern you describe or using `lib.concatLines` to now using `helpers.concatNonEmptyLines` _without_ a trailing newline.

Since the trailing newline is only added when there is content, I'm not sure I see the issue? Maybe I'm missing your point?

We could provide a `helpers.concatNonEmptyLines'` that takes an additional "trailing" boolean param, but this is a little over-engineered IMO.

File: tests/test-sources/modules/output.nix
Comment (MattSturgeon) [reply tests/test-sources/modules/output.nix:4]: > Is it intended that I can't just use module as a function without module attr?

It's not explicitly intended, it's just an edge-case I decided not to handle in order to keep the implementation simpler.

We could gate the `removeAttrs` behind a `isAttrs` or `? tests` check, but doing so felt a little bloated when I initially added full module support to the tests. :shrug: 

Don't really mind either way.

File: lib/utils.nix
Comment (stasjok) [reply lib/utils.nix]: > I'm not sure I see the issue

Since resulted config is passed through stylua, it's not a big deal. As I mentioned in comments several days ago, I often nit-picking those little things, so that my generated config would be nearly perfect. There is another case I didn't touch here, that I handled in my config: when doing `wrapDo` you actually need to indent every line (except empty lines). Because it's not handled now I intentionally wasn't indenting a first line of the string in `wrapVimscriptForLua`.

> since many of the places changed in _this_ PR are going from either using the pattern you describe or using `lib.concatLines` to now using `helpers.concatNonEmptyLines` _without_ a trailing newline.

What are many places? There is only one place where I changed `concatLines` to `helpers.concatNonEmptyLines`. It's in the opts module. And my version doesn't have problems, because those strings, that are concatenated by `concatNonEmptyLines` all has trailing newlines by itself. But old version has two trailing newlines (one from indented string and one from cancatLines).

For best results all strings concatenated by `concatNonEmptyLines` should have traling newlines. We can handle strings not having trailing newlines by ensuring all parts are have them. `removeSuffix "\n" + "\n"` should do it.

Comment (MattSturgeon) [reply lib/utils.nix]: A more general function that takes `sep` as an argument probably shouldn't be normalising line endings, since such functions are not always used on _lines_.

BTW the mapping function `(s: lib.removeSuffix "\n" s + "\n")` will always insert trailing newlines, like `lib.concatLines` does. It's just less likely to insert additional newlines if it is called with "dirty" input.

I think we should avoid our functions being too magic in unexpected ways. IMO the simplest implementation is to match upstream functionality, in this case just use `lib.concatLines` but filter the list for non-empty lines.

If we have an issue of passing in unintended `\n` in the input lines then we should address that in the calling code (could be another PR).

> It is similar to `concatLines` now, but with separator.

I don't see why we would ever need a different separator to `\n` if we know we're concatenating lines?

IMO this is getting overly complex for the sake of additional newlines that a) aren't that important for _this_ PR and b) should really be addressed in the calling code instead of the helper function.

Comment (MattSturgeon) [reply lib/utils.nix]: > Can you be more specific? What I need to do?

I think we should have:

```nix
concatNonEmptyLines = lines: concatLines (filter hasContent lines);
```

> Because I don't see a problem with my function. I never use `concatLines` myself, because the behavior I want almost always is `concatStringsSep "\n"` with a filter for empty strings.

If you really want, we could also have:

```nix
concatNonEmptyLines' =
  trailing: lines:
  let
    lines' = filter hasContent lines;
  in
  optionalString (lines' != [ ]) (
    concatStringsSep "\n" lines'
    + optionalString trailing "\n"
  );

concatNonEmptyLines = concatNonEmptyLines' true;
```

Depending on how often we really need that.

Comment (stasjok) [reply lib/utils.nix]: > b) should really be addressed in the calling code instead of the helper function.

So if I want it to be addressed _in the calling code_, do you recommend me to use this, righ?

```nix
    content =
      if config.type == "lua" then
        # Lua
        helpers.concatNonEmptyLines [
          config.extraConfigLuaPre
          (helpers.wrapVimscriptForLua config.extraConfigVim)
          config.extraConfigLua
          (lib.removeSuffix "\n" config.extraConfigLuaPost)
        ]
      else
        # Vimscript
        helpers.concatNonEmptyLines [
          (helpers.wrapLuaForVimscript config.extraConfigLuaPre)
          config.extraConfigVim
          (lib.removeSuffix "\n" (
            helpers.wrapLuaForVimscript (
              helpers.concatNonEmptyLines [
                config.extraConfigLua
                (lib.removeSuffix "\n" config.extraConfigLuaPost)
              ]
            )
          ))
        ];
```

Comment (stasjok) [reply lib/utils.nix]: Anyway, you are a maintainer here, so I'll change as you suggested even though your suggested function definitely not want I need. But anyway, let's it have extra newlines. I don't have enough energy to argue, better to finish it somehow, because I guess we won't agree on this one.

Comment (MattSturgeon) [reply lib/utils.nix]: If `extraConfigLuaPost` is likely to have trailing newline(s) then so is `extraConfigLua`, `extraConfigVim` and `extraConfigLuaPre`.

If we're getting that nit-picky, then `removeSuffix "\n"` isn't a robust solution, since there could be any number of trailing newlines hypothetically.

IMHO this isn't in the scope of this PR and we aren't making things any worse than the existing implementation by going with the simple option for now.

If you really think it's important to normalise the ending of those options, maybe we can do that in the options' `apply` functions? But again, I think we should keep that out of _this_ PR.

Comment (MattSturgeon) [reply lib/utils.nix]: > I'll change as you suggested even though your suggested function definitely not what I need. But anyway, let's it have extra newlines. I don't have enough energy to argue

Sorry if that got unnecessarily heated. That wasn't my intention! :sweat_smile: 

I'm not saying we definitely shouldn't deal with redundant newlines, I just don't want to deal with that additional complexity in this PR since it primarily focuses on keeping the output empty by default.

I'm also not saying we definitely should fix them, but we can discuss potential solutions in another issue/discussion/PR/matrix.

Comment (stasjok) [reply lib/utils.nix]: > If extraConfigLuaPost is likely to have trailing newline(s) then so is extraConfigLua, extraConfigVim and extraConfigLuaPre.

But my problem is with with `concatNonEmptyLines` producing unnecessarily newline at the end, so I don't need to strip other variables. They will be concatenated properly. Only the last one shouldn't have newline to count the one added by concatLines. So my complain is that you take my function that get the job done (because my guess is that at least 90% strings already have trailing newlines due to using indented string style) and turn in the function that doesn't.

> If we're getting that nit-picky, then `removeSuffix "\n"` isn't a robust solution, since there could be any number of trailing newlines hypothetically.

That's right, but caller may want it to have some trailing newlines, that's why I usually strip only one. Usecase for this is that both strings from here will be similar:

```nix
{
  a = "txt;
  b = ''
    txt
  ''
}
```

> IMHO this isn't in the scope of this PR and we aren't making things any worse than the existing implementation by going with the simple option for now.

Originally not, but when you suggested improving, I take it to my heart. So with my second commit I wanted to not only produce _whitespace only_ config, but clean up config generation. Anyway, I already gave in to do it your way. I guess comments are refreshed too late, because we often reply to the penultimate comments.

File: modules/opts.nix
Comment (MattSturgeon) [inline modules/opts.nix:87]: Is there a scenario where `content` could be empty whitespace, but not actually `""`?

I think `concatNonEmptyLines` ensures that its result either has content or is `""`.

(`concatLines` only has a trailing newline if the list is non-empty, and the list will be empty if all items failed the `hasContent` filter).

Not an issue, I'm just asking for my own understanding... In case I'm missing something again :grin:

**Review by MattSturgeon — APPROVED** [review]
LGTM, thanks for your efforts and your patience explaining things to me and debating various pros/cons!

I think this change will be appreciated by many!

@nix-community/nixvim any final suggestions before merging?

File: modules/opts.nix
Comment (stasjok) [reply modules/opts.nix:87]: Your are right. My mistake. Mistake was made because because earlier I considered two options:

1. `concatLines` at the top and `helpers.hasContent content` as mkIf condition.
2. `helpers.concatNonEmptyLines` at the top and `content != ""` as a mkIf condition.

So I was rushed changing it before dinner and thought that since I'm adding something with `concatLines` at the top, I have to change also the condition, but I actually don't. With _NonEmpty_ part of the function it would work fine.

Comment (stasjok) [reply modules/opts.nix]: Changed it back.

Comment (MattSturgeon) [reply modules/opts.nix]: No problem!

I hope you don't feel too much pressure to produce perfect code; there's really no such thing. :joy:

Everything can always be improved (or done differently) and sometimes it just comes down to preference or giving up and "just making it work".

Feedback in a code review is never personal and isn't always meant as criticism :smiley:

Comment (stasjok) [reply modules/opts.nix]: > Everything can always be improved (or done differently) and sometimes it just comes down to preference or giving up and "just making it work".

In this case I don't agree that this is just "doing differently". For example I don't have complains in the byte-compilation PR, because I can understand that my code could be not very bad (actually I'm not a programmer, I'm sysops, and I'm junior at best at programming), or suitable, or maintainable. In this case I just think that the new behavior is wrong and counter-interact what I already fixed.

```
nix-repl> c.config.content  
"-- Ignore the user lua configuration\nvim.opt.runtimepath:remove(vim.fn.stdpath('config'))              -- ~/.config/n
vim\nvim.opt.runtimepath:remove(vim.fn.stdpath('config') .. \"/after\")  -- ~/.config/nvim/after\nvim.opt.runtimepath:r
emove(vim.fn.stdpath('data') .. \"/site\")     -- ~/.local/share/nvim/site\n\n"
```

See "\n\n" at the end? They will appear in most cases where this function is used. Fortunately stylua will strip all extra newlines, so I'm fine with that at the end. Just need to relax :).

Comment (MattSturgeon) [reply modules/opts.nix]: > `vim.opt.runtimepath:remove(vim.fn.stdpath('config'))              -- ~/.config/nvim`

Is this correct output? At a glance it looks like it's missing a linebreak :thinking: 

> In this case I don't agree that this is just "doing differently".
> See "\n\n" at the end? They will appear in most cases where this function is used.

As I said before, we can revisit in the future and figure out the best solution.

We don't need to achieve perfection in a single PR :grinning: 

> Fortunately stylua will strip all extra newlines, so I'm fine with that at the end. Just need to relax :)

:heart:

Comment (stasjok) [reply modules/opts.nix]: > Is this correct output? At a glance it looks like it's missing a linebreak 🤔

Everything is fine except trailing newlines. It's added to every wrapped config here:

https://github.com/nix-community/nixvim/blob/6dc0bda459bcfb2a38cf7b6ed1d6a5d6a8105f00/modules/top-level/output.nix#L150-L153

Comment (MattSturgeon) [reply modules/opts.nix]: I think if you had additional "lines" in the `concatNonEmptyLines` call, they would also (usually) end up with `"\n\n"`, one `\n` from the line's definition and one from the concatenation.

While this would likely make the string more readable, you could debate whether or not it is "intended". So it's not just the _final_ trailing line that is an issue, although that is _more_ of an issue.

I think in another PR we can look into normalising this and figuring out exactly what we want. Unless we decide to just rely on stylua for formatting issues like this.

Comment (stasjok) [reply modules/opts.nix]: > I think if you had additional "lines" in the `concatNonEmptyLines` call, they would also (usually) end up with `"\n\n"`, one `\n` from the line's definition and one from the concatenation.

Are you talking about the middle strings? That's exactly what I want: so that every string (I call it config part) is separated with empty line (empty line means double newline characters). So it is exactly what I want: double newlines between config parts and a single newline at the end. My original function was just relying on provided strings already have trailing newlines (which is the case for almost all config parts, because they are written with indented strings). So my original function worked good enough so that I didn't bother much, or would want to fix original string definition to add trailing newline for consistent behavior. So the only thing that I possibly may want to impoove is to normalize strings. Look again at my example of my «perfect» function:

```
nix-repl> lib.x86_64-linux.helpers.concatNonEmptyLinesSep "" ["a" "b\n" "" "\n" "c"]   
"a\nb\nc\n"

nix-repl> lib.x86_64-linux.helpers.concatNonEmptyLinesSep "\n" ["a" "b\n" "" "\n" "c"] 
"a\n\nb\n\nc\n"
```

Second case is what I wanted in modules/output: two newlines between config parts and only one at the end.

> I think in another PR we can look into normalising this and figuring out exactly what we want. Unless we decide to just rely on stylua for formatting issues like this.

Let's not debate on this anymore. I'm not planning making another PR. The implementation here was good enough. All that can be improved from my point of view is adding my _too magic_ function. I think it's exactly what I want most of the time and it is suitable for both cases: when I want double newlines between strings and when one newline is enough.

Stylua is great, it'll fix all trailing newlines and chomp all triple, quadruple etc newlines to double. It doesn't do anything for Vim-type configs, but who is even using it?

**Review by GaetanLepage — APPROVED** [review]
LGTM after a rapid glance.
I trust @MattSturgeon for the details.

Thank you very much @stasjok for this great work and for addressing the detailed review !

**Comment by GaetanLepage** [comment]
@mergifyio queue

**Pull request merged**
